src/models/geograph.py

```diff
@@ -43,20 +123,179 @@ def graph(self) -> nx.Graph:
+            but Shapefiles also work.
+            array (np.ndarray, optional): 2D numpy array with the raster data
+            that can be passed as an alternative to `raster_path`.
+            mask (np.ndarray, optional): Boolean mask that can be applied over
+            the polygonisation. Defaults to None.
+            transform (affine.Affine, optional): Affine transformation to apply
+            when polygonising. Defaults to the identity transform.
+            crs (str, optional): Coordinate reference system to set on the
+            resulting dataframe. Defaults to None.
+
+        Raises:
+            ValueError: If `save_path` is not a GPKG or Shapefile, or if both
+            `raster_path` and `array` are None.
+
+        Returns:
+            STRtree: The Rtree object used to build the graph.
         """
-        Load networkx graph object from a pickle file.
+        if save_path.suffix not in (".shp", ".gpkg"):
+            raise ValueError("Argument `save_path` should be a GPKG or Shapefile.")
+        if array is None and raster_path is None:
+            raise ValueError("One of `raster_path` or `array` arguments must be used.")
+        elif array is None and raster_path is not None:
+            with rasterio.Env():
+                with rasterio.open(raster_path) as image:
+                    data = image.read(1)  # Read band 1
+        else:
+            data = array
+        vector_df = polygonise(
+            data_array=data,
+            mask=mask,
+            connectivity=4,
+            apply_buffer=True,
+            transform=transform,
+            crs=crs,
+        )
+
+        if save_path.suffix == ".gpkg":
+            vector_df.to_file(save_path, driver="GPKG")
+        else:
+            vector_df.to_file(save_path)
+        return self._dataframe_to_graph(
+            vector_df, attributes=["geometry", "class_label"], tolerance=self.tolerance
+        )
+
+    def _load_graph(self, graph_path: pathlib.Path) -> STRtree:
+        """
+        Load networkx graph and rtree objects from a pickle file.
 
         Args:
-            graph_path (str): Path to a pickle file.
+            graph_path (pathlib.Path): Path to a pickle file. Can be compressed
+            with gzip or bz2.
+
+        Raises:
+            ValueError: If `graph_path` is not a pickle, bz2, or gz file.
+
+        Returns:
+            STRtree: The Rtree object used to build the graph.
         """
-        if not graph_path.endswith(("pickle", "pkl")):
+        if graph_path.suffix not in (".pickle", ".pkl", ".gz", ".bz2"):
             raise ValueError("Argument `graph_path` should be a pickle file.")
-        self.graph = nx.read_gpickle(graph_path)
+
+        if graph_path.suffix == ".bz2":
+            bz2_file = bz2.BZ2File(graph_path, "rb")
+            data = pickle.load(bz2_file)
+            bz2_file.close()
+        elif graph_path.suffix == ".gz":
+            gz_file = gzip.GzipFile(graph_path, "rb")
```

Comment:
nit: not really needed since it's so short that the close comes immediately after, but maybe we want to use context `with ... ` here? (up to you, this is really nitpicky but just crossed my mind)

Reply:
I originally had issues with using `with` for some reason, but I tried it again and it seems to work so I've changed it.